core/src/main/java/org/apache/iceberg/DataFiles.java

```diff
@@ -334,4 +338,10 @@ public DataFile build() {
           keyMetadata, splitOffsets);
     }
   }
+
+  private static <E> List<E> copyList(List<E> toCopy) {
+    List<E> copy = Lists.newArrayListWithExpectedSize(toCopy.size());
+    copy.addAll(toCopy);
+    return copy;
```

Comment:
`GenericDataFile` also wraps the list into `Collections.unmodifiableList`. Does it make sense here?

Reply:
I don't think it is necessary. In the long term, we should move these back to using ImmutableList. Right now, we just need to fix the serialization problem. There are other mutable instances being passed in from Metrics as well. We can fix all that later, but unblock the release now.